src/librustc_mir/transform/remove_unit_storage.rs

```diff
@@ -0,0 +1,29 @@
+//! The `RemoveUnitStorage` pass removes `StorageLive` and `StorageDead` statements
+//! which operates on locals of type `()`.
+
+use crate::transform::{MirPass, MirSource};
+use rustc::mir::*;
+use rustc::ty::TyCtxt;
+use smallvec::SmallVec;
+
+pub struct RemoveUnitStorage;
+
+impl<'tcx> MirPass<'tcx> for RemoveUnitStorage {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, _: MirSource<'tcx>, body: &mut BodyAndCache<'tcx>) {
+        let unit_locals: SmallVec<[_; 32]> =
+            body.local_decls.iter().map(|local| local.ty == tcx.types.unit).collect();
```

Comment:
I'm far from an expert in this code, but what's the need to allocate an array here instead of just indexing into `body.local_decls` and comparing the types on line 20?

Reply:
the code below mutates the body, so we can't hold an immutable at the same time. technically it is doable, but not with the current API. Maybe `basic_block_mut` could return a reference to the rest of the body... not sure, but there's still quite some design space available here

Reply:
Shouldn't you be able to use https://doc.rust-lang.org/nightly/nightly-rustc/rustc/mir/struct.BodyAndCache.html#method.basic_blocks_and_local_decls_mut for that?